src/main/java/com/google/jenkins/plugins/computeengine/ssh/GooglePrivateKey.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.jenkins.plugins.computeengine.ssh;
+
+import com.google.jenkins.plugins.computeengine.SshConfiguration;
+import hudson.util.Secret;
+import java.io.Serializable;
+
+/** Class to store optional custom private key selected by user */
+public class GooglePrivateKey extends GoogleKeyCredential implements Serializable {
+  private final Secret privateKey;
+
+  private GooglePrivateKey(Secret privateKey, String user) {
+    super(user);
+    this.privateKey = privateKey;
+  }
+
+  public static GooglePrivateKey generate(String credentialsId, String user) {
+    String privateKeyStr =
+        SshConfiguration.getCustomPrivateKeyCredentials(credentialsId).getPrivateKeys().get(0);
+    return new GooglePrivateKey(Secret.fromString(privateKeyStr), user);
+  }
+
+  public Secret getPrivateKey() {
+    return privateKey;
+  }
+
+  @Override
+  public String toString() {
+    return "Private key:\n" + privateKey.getEncryptedValue();
```

Comment:
Okay, security question.  So, we've got this private key available to the overall Jenkins process.   Should we assume that any Jenkins build job can now grab this private key and do anything they want with it?  Could someone, for example, write a change to a build job and check it in that grabs the private key and sends it to a remote server that then uses it maliciously?

Reply:
I don't believe that this should be a problem. If this is a problem, then it would be a security issue with the GoogleKeyPair class as well. It seems like the GoogleKeyPair class originally stored the private key as a String and it got changed to a Secret object. I modeled the GooglePrivateKey class directly based off of the GoogleKeyPair class, as in [this commit](https://github.com/jenkinsci/google-compute-engine-plugin/commit/16d2ae71a1b34c81db1d74f83c41577536e5256f#diff-181efb7128a369621fe4258b01509ba2819a393cda023dea411a700680161f38), in order to meet security standard.

Reply:
That goes with any single credential stored in Jenkins. That is always a risk.